azure/services/publicips/client.go

```diff
@@ -18,18 +18,25 @@ package publicips
 
 import (
 	"context"
+	"encoding/json"
 
 	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
 	"github.com/Azure/go-autorest/autorest"
+	azureautorest "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
 // Client wraps go-sdk.
 type Client interface {
-	Get(context.Context, string, string) (network.PublicIPAddress, error)
-	CreateOrUpdate(context.Context, string, string, network.PublicIPAddress) error
-	Delete(context.Context, string, string) error
+	Get(context.Context, azure.ResourceSpecGetter) (result interface{}, err error)
+	CreateOrUpdateAsync(context.Context, azure.ResourceSpecGetter, interface{}) (result interface{}, future azureautorest.FutureAPI, err error)
+	DeleteAsync(context.Context, azure.ResourceSpecGetter) (future azureautorest.FutureAPI, err error)
+	IsDone(context.Context, azureautorest.FutureAPI) (isDone bool, err error)
+	Result(context.Context, azureautorest.FutureAPI, string) (result interface{}, err error)
 }
 
 // AzureClient contains the Azure go-sdk Client.
```

Comment:
If possible we should try to make this type private. AFAICT it's being used in the `virtualmachines` package but now that we've refactored the `Get()` to implement the `async.Getter` interface you should be able to use the `async.Getter` interface as type for the Getter just like we're already doing for network interfaces https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2317/files#diff-bd4c935d5db79644e733f4f321bda2011252db8da4cc4bdfef5fffe3e5cd0710L55

Reply:
Are you able to make this type private (unexported) now that you removed usage in the virtualmachines package?

Reply:
My bad, you are correct. It still needs to be a public type since we're returning it in the NewClient func.